include/envoy/server/filter_config.h

```diff
@@ -90,11 +90,12 @@ class ProtocolOptionsFactory : public Config::TypedFactory {
    * @param validation_visitor message validation visitor instance.
    * @return Upstream::ProtocolOptionsConfigConstSharedPtr the protocol options
    */
-  virtual Upstream::ProtocolOptionsConfigConstSharedPtr
-  createProtocolOptionsConfig(const Protobuf::Message& config,
-                              ProtobufMessage::ValidationVisitor& validation_visitor) {
+  virtual Upstream::ProtocolOptionsConfigConstSharedPtr createProtocolOptionsConfig(
+      const Protobuf::Message& config,
+      Server::Configuration::TransportSocketFactoryContext& factory_context) {
     UNREFERENCED_PARAMETER(config);
-    UNREFERENCED_PARAMETER(validation_visitor);
+    UNREFERENCED_PARAMETER(factory_context);
+    throw EnvoyException("not implemented");
```

Comment:
Why this alwasy throw? We return nullptr to make protocol options optional, so this shouldn't throw not implemented exception. If we're going to require all protocol options to implement this, then just make this method PURE.

Reply:
we do not make it PURE as NamedNetworkFilterConfigFactory,NamedHttpFilterConfigFactory, and  NamedUpstreamNetworkFilterConfigFactory inherit from it and benefit from a default impl.

see more info about why I added the exception in the discussion above:
https://github.com/envoyproxy/envoy/pull/12234#discussion_r459848281

happy to change it once we reach concencus

Reply:
Make override there to return nullptr, or throw exceptions and here PURE?

I think we explicitly allow this to be nullptr,
https://github.com/envoyproxy/envoy/blob/68d74ade909e1eb757da674f10ba73ccad7f93c2/source/common/upstream/upstream_impl.cc#L191 , changing the default here scares me and might cause unexpected NACK from config. Not sure why there's no test caught this though.

Reply:
If we want fix this I'd suggest in another PR since it's not related to the refactor/plumb.

Reply:
if I understand you correctly, you suggest to make this method PURE, and add implementations that throw in the sub-classes?

Reply:
No, I'm saying we shouldn't change this one here. That should be in a separate PR anyway.

The difference between PURE vs NOT_IMPLEMENTED (if you throw use macro) is failing at build time or runtime. If it is not supposed to be called it should be failing at build time.

Reply:
i can revert this change. cc @kyessenov